### ISSUE TITLE
Ad Collecting: Google Advert

### ISSUE THREAD

**Issue by cqx931** [dhowe/AdNauseam#1555]
Chrome:
Right now most google ads are hidden from this filter from Easylist: `.adsbygoogle`. However, from my observation, hiding this element is going to prevent the whole ad from rendering, which is not what we want for adnauseam. 

Currently, to collect ad on the testing website for experiment, I'm unhiding this rule and hide an inner element `#google_image_div` instead. AdNauseam can successfully collect the ad from this solution, however, it's kind of annoying that a blank grey rectangle is going to be left on the website and this is `.adsbygoogle`. 

Is it better to hide the blank grey rectangle and not being able to click most of the google ads or displaying the grey rectangle and offering the possibility to click the ad? Any other workaround here?

Procedural cosmetic filters might be helpful: https://github.com/gorhill/uBlock/wiki/Procedural-cosmetic-filters

Firefox（Fixed now）
Collecting might not work because this blocking rule triggered by Easylist. `/googleads.$domain=~googleads.github.io`

See https://github.com/dhowe/AdNauseam/issues/1364

**Comment by cqx931** [comment]
<img width="848" alt="Screen Shot 2019-09-17 at 10 52 50 AM" src="https://user-images.githubusercontent.com/3056154/65053051-5beaf980-d939-11e9-9c7b-161a706d0ba3.png">

**Comment by dhowe** [comment]
Its super important that we are able to collect (and then click) google ads -- these make up the vast majority of ads out there. Can we set up a very simple page that we can use to test solutions for this?

**Comment by cqx931** [comment]
On the front page of the new website, there is a simple google display ad showing there: http://filbert.cc/

I also put the same ad code to a blank page. The ad displayed there(at least viewed from my end) is usually a mix of text and images.
http://filbert.cc/test.html

**Comment by dhowe** [comment]
@cqx931 I think repeatedly hitting that site would be interpreted as click-fraud and potentially get us banned very quickly

**Comment by cqx931** [comment]
yes, but just to test ad collecting without clicking won’t get us banned, which is how I test it at the moment.

**Comment by cqx931** [comment]
@dhowe I think it makes most sense to have a test case with valid google adsense ad instead of a static version with similar dom structure.
One argument will be a static dom structure can't reproduce the issue we have with google ad this time, because it is involving blocking rules from google.

We can create test cases with similar dom structure to solve a particular ad collection case. But for testing and being aware of the latest changes in Google Adsense, it is more helpful to have a real case.

Finding another site with simple google ad sense ad display is also another solution. But then the ad might disappear at any time if the site owner removes it. Also I think no matter what we test for ad collection, disable ad clicking for testing purpose should be a good practice.

**Comment by cqx931** [comment]
Update. @dhowe To progress on this, I think there are two stages that need to be done for this ticket:

1. To be able to collect this typical google image ad in general: We want a generic unhiding rule for `.adsbygoogle`. 
Currently I can only unhide certain rules for specific domain. Since we are not going to be able to delete this rule from easylist, the solution I can think about is to ignore that specific rule in the cosmetic filter parsing process.
Then the generic hiding rule in adnauseam.txt: `###google_image_div` is going to pick up the ad if it is an image ad, or other hiding rules if needed for dynamic ads.
*If this is necessary, I think we should include it in the next new release.

2.Fix the grey rectangle issue.
Stage 1 can make sure we can collect google image ads but not visually perfect. The user is still able to see this grey rectangle from `.adsbygoogle`. To hide it, one solution would be to change the div height to 0. However, to my knowledge, this is also not supported by the current mechanism in ublock origin. Procedural cosmetic filters are designed for css based selector, but not change it. Also, procedural cosmetic filters are also designed for non-generic rules.

**Comment by dhowe** [comment]
> 1. To be able to collect this typical google image ad in general: We want a generic unhiding rule for `.adsbygoogle`. 
Currently I can only unhide certain rules for specific domain. Since we are not going to be able to delete this rule from easylist, the solution I can think about is to ignore that specific rule in the cosmetic filter parsing process.

Are you suggesting we handle that case specifically in adnauseam code?

**Comment by cqx931** [comment]
Or we can make generic exception rules working only for adnauseam filter, by changing how unlock origin handle exception rule for adnauseam filter.

Sally

**Comment by dhowe** [comment]
Interesting suggestion -- since this policy rule would be applied only to our own list, we could avoid the problem of an bad actor sneaking an exception rule into a crowd-sourced list and unblocking all of their own ads.

However we'd still have to be careful about properly vetting any generic exception rules in adnauseam.txt.

But I far prefer this to making special cases in code for google

**Comment by cqx931** [comment]
Good. I opened another ticket for this function only. See: https://github.com/dhowe/AdNauseam/issues/1561

**Comment by cqx931** [comment]
After checking the code I realized that ublock actually implemented generic unhiding rule despite the fact that it is not mentioned in their wiki. 
Here is the code: https://github.com/dhowe/AdNauseam/blob/master/src/js/cosmetic-filtering.js#L681

Double checked that generic unhiding rule from adnauseam does get parsed correctly in the back ground:
![image](https://user-images.githubusercontent.com/3056154/65640737-0ce83880-dfb9-11e9-874e-6dba1f4f2d88.png)

So I updated the unhiding rule from domain specific to generic in the filter:
https://github.com/dhowe/AdNauseam/blob/master/filters/adnauseam.txt#L20

**Comment by cqx931** [comment]
So Stage 1 is fixed and now is Stage 2: the grey rectangle.

One thing I noticed about the grey rectangle is that I can only see the grey rectangles from our testing site (live google adsense), but not on my clipped google ad samples page.
http://rednoise.org/adntest/googleAdSenseCases.html

However I still think the best practice here is to find a solution to change the height of the div, as we still don't like weird blank spaces on the web...Meanwhile, another solution would be to hide `.adsbygoogle` after unhiding them at the first place. Any thoughts on this part? @dhowe

**Comment by dhowe** [comment]
is there a way to schedule things sequentially (I don't think so), so that the .adsbygoogle div is allowed, processed, then hidden later?

**Comment by cqx931** [comment]
No, not with pure filters. There are different phases for different kinds
of cosmetic filters to be applied in ublock. We can look into the code here
and see whether we can create a new type of cosmetic filter and reapply
them at later stages.

**Comment by cqx931** [comment]
See https://github.com/dhowe/AdNauseam/issues/1556

**Comment by chirag127** [comment]
> @dhowe I think it makes most sense to have a test case with valid google adsense ad instead of a static version with similar dom structure. One argument will be a static dom structure can't reproduce the issue we have with google ad this time, because it is involving js script from google's side as well as blocking rules for google.
> 
> We can create test cases with similar dom structure to solve a particular ad collection case. But for testing and being aware of the latest changes in Google Adsense, it is more helpful to have a real case.
> 
> Finding another site with simple google ad sense ad display is also another solution. But then the ad might disappear at any time if the site owner removes it. Also I think no matter what we test for ad collection, disable ad clicking for testing purpose should be a good practice.

Oh yes did someone set up an AdSense account to check if ad nauseam works or not or all our clicks are going towards the hole which is called click fraud or invalid clicks

**Comment by dhowe** [comment]
https://rednoise.org/AdNauseamVsGoogle.pdf

**Comment by chirag127** [comment]
> https://rednoise.org/AdNauseamVsGoogle.pdf

I was written last year, I am not sure the findings of the research still apply after a year.

I think we should test it atleast once a year.

**Comment by dhowe** [comment]
> @dhowe I think it makes most sense to have a test case with valid google adsense ad instead of a static version with similar dom structure. 

makes sense to me
